server-function/src/index.ts

```diff
@@ -0,0 +1,106 @@
+import path from 'path';
+import fs from 'mz/fs';
+import mimeTypes from 'mime-types';
+import ms from 'milliseconds';
+import fresh from 'fresh';
+
+// create-react-app uses only 32 bits for hashes, allow caching for a limited
+// time for lower chance of collision
+const CACHE_TIME_SEC = ms.months(1) / 1000;
+
+export async function isFile(filePath: string): Promise<fs.Stats | undefined> {
+  try {
+    const stats = await fs.stat(filePath);
+    return stats.isFile() ? stats : undefined;
+  } catch (err) {
+    if (err.code === 'ENOENT') {
+      return undefined;
+    }
+    throw err;
+  }
+}
+
+interface ServeFile {
+  action: 'serveFile';
+  fullPath: string;
+  contentType: string | false;
+  cacheHint: any;
+  status?: number;
+}
+
+interface SendStatus {
+  action: 'sendStatus';
+  status: number;
+}
+
+interface Invoke {
+  action: 'handleInvoke';
+}
+
+type Decision = ServeFile | Invoke | SendStatus;
+
+export class Server {
+  constructor(
+    private directory: string,
+    private extensions: string[] = ['.html'],
+    private cachedPath = '/static',
+  ) {
+  }
+  public async handle(url: string, headers: { [k: string]: string | string[] | undefined }): Promise<Decision> {
```

Comment:
This makes me think that we should have just made a custom runtime for shift functions implemented with express.
Would have been more robust, used standard tools and would have been easier to implement and test.

Reply:
Mostly agree, the cost this API is not in migration but breaking code reuse which assumes a framework.